merkle-tree/bounded-vec/src/lib.rs

```diff
@@ -335,7 +336,8 @@ where
         Self {
             capacity,
             length: 0,
-            next_index: 0,
+            first_index: 0,
+            last_index: 0,
```

Comment:
why do we need a first and a last index?
Don't we always start from index 0 except when providing a different index to start from?

Reply:
We used to do that, but I think it's not a correct thing to do.

When we don't provide a different index, I think we should start from the **oldest** element. The oldest element can be somewhere in the middle, if we went over capacity.

Let's assume that we have a bounded vec of integers incremented by 1. The capacity is 16. We append 22 elements. It ends up looking like:

```
                              $  ^
[16, 17, 18, 20, 21, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]

^ - first_index
$ - last_index
```

Every time we are over capacity and add more elements, we need to increment `first_index` and `last_index`.

In my opinion, a default iterator should iterate from 5 to 21.

What you are proposing, would iterate... how exactly? From 16 to 21? From 16 to 15?
We used to iterate from 16 to 15, which I think it's incorrect. This approach made the previous implementation of `.to_slice()` a bit useless and it was the reason why we ended up with doing all this manual iteration dance in `ConcurrentMerkleTree`.

Maybe I should rename `first_index` and `last_index` to `head` and `tail` to make it clearer?

Reply:
your example makes it clear what it means

Reply:
last index is always first index minus 1 except if first index = 0 then its length - 1, and it is 0 if it is empty.
So what is the reasoning to put it into a variable and not just compute it when needed?

Reply:
fair I can see that it makes sense for the iterator to start at the current index

Reply:
wrt heads and tail it's confusing in a different way, I think we just need a doc comment then it's clear.